[DataGridPro] Add support for master/detail (#3387)
mui-org/material-ui-x: docs/src/pages/components/data-grid/group-pivot/group-pivot.md

{{"demo": "pages/components/data-grid/group-pivot/TreeDataFullExample.js", "bg": "inline", "defaultCodeOpen": false}}

> ## Master detail [<span class="plan-pro"></span>](https://mui.com/store/items/material-ui-pro/)

Reviewer: You can also remove the 🚧 from the [feature comparison](https://deploy-preview-3387--material-ui-x.netlify.app/components/data-grid/getting-started/#feature-comparison)

For [Pricing page](https://mui.com/pricing/) I don't know if it is on X or core repo

Author: To update the Pricing page it needs to be another PR in the core repo.

{{"demo": "pages/components/data-grid/group-pivot/TreeDataFullExample.js", "bg": "inline", "defaultCodeOpen": false}}

## Master detail [<span class="plan-pro"></span>](https://mui.com/store/items/material-ui-pro/)

> The master detail allows to expand a row to display additional information inside a panel.

Reviewer 2: ```suggestion
The master detail feature allows expanding a row to display additional information inside a panel.
```

> To start using this feature, pass a function to the `getDetailPanelContent` prop with the content to be rendered inside the panel.

Reviewer 2: ```suggestion
To use this feature, pass a function to the `getDetailPanelContent` prop with the content to be rendered inside the panel.
```

Any valid React element can be used as the row detail, even another grid.

The height of the detail panel content needs to be provided upfront.

> The grid assumes the value of 500px by default.

Reviewer 2: ```suggestion
The grid assumes the value of 500px by default
```

> However, this can be configured by passing a function to the `getDetailPanelHeight` prop.

Reviewer 2: ```suggestion
however this can be configured by passing a function to the `getDetailPanelHeight` prop that returns the required height.
```

> Both props are called with a [`GridRowParams`](/api/data-grid/grid-row-params/) object, allowing to return a different value for each row.

Reviewer 2: ```suggestion
Both props are called with a [`GridRowParams`](/api/data-grid/grid-row-params/) object, allowing you to return a different value for each row.
```

```tsx
<DataGridPro
  getDetailPanelContent={({ row }) => <div>Row ID: {row.id}</div>}
  getDetailPanelHeight={({ row }) => 100} // Optional, default is 500px.
/>
```

> To expand a row, click in the "i" icon or press <kbd><kbd class="key">Ctrl</kbd>+<kbd class="key">Enter</kbd></kbd> inside one of the cells of the row.

Reviewer 2: ```suggestion
To expand a row click on the "i" icon, or press <kbd><kbd class="key">Ctrl</kbd>+<kbd class="key">Enter</kbd></kbd> inside one of the cells of the row.
```

Returning `null` or `undefined` as the value of `getDetailPanelContent` will prevent the respective row from being expanded.

{{"demo": "pages/components/data-grid/group-pivot/BasicDetailPanels.js", "bg": "inline", "defaultCodeOpen": false}}

> ⚠ Always memoize the function provided to `getDetailPanelContent` and `getDetailPanelHeight`.

> > The grid bases on the referential value of these props to cache their values and optimize the rendering.

Reviewer 2: ```suggestion
> The grid depends on the referential value of these props to cache their values and optimize the rendering.
```

>
> ```tsx
> const getDetailPanelContent = React.useCallback(() => { ... }, []);
>
> <DataGridPro getDetailPanelContent={getDetailPanelContent} />
> ```
>
> ⚠ Depending on the height of the detail panel, you may see a blank space when scrolling.

> > This is caused because the grid uses a lazy approach to update the rendered rows.

Reviewer 2: ```suggestion
> This is because the grid uses a lazy approach to update the rendered rows.
```
or
```suggestion
> This is caused by the grid using a lazy approach to update the rendered rows.
```

> > Set `rowThreshold` to 0 to fix it:

Reviewer 2: ```suggestion
> Set `rowThreshold` to 0 to increase the rendering speed at the cost of performance:
```
(please correct my assumption...)

Author: It's not very the "rendering speed". I added more information to this warning.

>
> ```tsx
> <DataGridPro rowThreshold={0} />
> ```

### Controlling the detail panels

> To control which rows are expanded, pass a list of row ids to the `detailPanelExpandedRowIds` prop.

Reviewer 2: ```suggestion
To control which rows are expanded, pass a list of row IDs to the `detailPanelExpandedRowIds` prop.
```
?

Author: "IDs" seem right.

> The `onDetailPanelExpandedRowIds` prop can be used to detect when a row is expanded or collapsed.

Reviewer 2: > can be used to detect when a row is expanded

"is expanded", or "is requesting to be expanded"? Is this a controlled component, or is this a callback prop? (Asking because of the preceding use of "control".)

Author: I added more context about what is `onDetailPanelExpandedRowIds`.

> On the other hand, if you only want to initialize the grid with some rows already expanded, use the `initialState` prop as follow:

Reviewer 2: Does this mean that they are expanded by default and can't be collapsed?

```tsx
<DataGridPro initialState={{ detailPanel: { expandedRowIds: [1, 2, 3] } }}>
```

{{"demo": "pages/components/data-grid/group-pivot/ControlMasterDetail.js", "bg": "inline", "defaultCodeOpen": false}}

> ### Using a form as detail panel

Reviewer 2: ```suggestion
### Using a detail panel as a form
```

> As an alternative to the built-in [row editing](/components/data-grid/editing/#row-editing), a form component can be rendered inside the detail panel, allowing to edit the current row values.

Reviewer 2: ```suggestion
As an alternative to the built-in [row editing](/components/data-grid/editing/#row-editing), a form component can be rendered inside the detail panel, allowing the user to edit the current row values.
```

> The following demo integrates with [react-hook-form](https://react-hook-form.com/), but other form libraries can also be used instead.

Reviewer 2: ```suggestion
The following demo shows integration with [react-hook-form](https://react-hook-form.com/), but other form libraries can also be used instead.
```

{{"demo": "pages/components/data-grid/group-pivot/FormDetailPanel.js", "bg": "inline", "defaultCodeOpen": false}}

> ### Customize the detail panel toggle

Reviewer 2: ```suggestion
### Customize the detail panel toggle
```
```suggestion
### Customizing the detail panel toggle
```

To change the icon used for the toggle, you can provide a different component for the [icon slot](/components/data-grid/components/#icons) as follow:

```tsx
<DataGridPro components={{ DetailPanelToggleIcon: CustomIcon }}>
```

> Although, if this is not sufficient, the entire toggle component can be overriden.

Reviewer 2: ```suggestion
If this is not sufficient, the entire toggle component can be overridden.
```

To fully customize it, define a column with `field: GRID_DETAIL_PANEL_TOGGLE_FIELD`.

> The new toggle component can be provided via `renderCell`.

Reviewer 2: ```suggestion
The new toggle component can be provided via `renderCell`.
```
There's some assumed knowledge here. Without looking at the example, one might assume that `renderCell` is a prop. Can we give some contextual hints as to what it is, and/or perhaps link to a relevant section of the docs? (it doesn't come up in Algolia search)

Author: I rephrased this entire paragraph.

> It is recommended to first spread `GRID_DETAIL_PANEL_TOGGLE_COL_DEF` when defining the column, since it provides a good set of default values to start with.

Reviewer 2: ```suggestion
It is recommended to first spread `GRID_DETAIL_PANEL_TOGGLE_COL_DEF` when defining the column, since it provides a good set of default values.
```

```tsx
<DataGridPro
  columns={[
    {
      ...GRID_DETAIL_PANEL_TOGGLE_COL_DEF, // Already contains the right field
      renderCell: (params) => <CustomDetailPanelToggle {...params}>
    },
  ]}
/>
```

> This approach can also be used to change the location of the toggle column, as showed below.

Reviewer 2: ```suggestion
This approach can also be used to change the location of the toggle column, as shown below.
```

> The grid detects the special field and does not add a second column.

Reviewer 2: Special field?

{{"demo": "pages/components/data-grid/group-pivot/CustomizeDetailPanelToggle.js", "bg": "inline", "defaultCodeOpen": false}}

> **Note**: As any ordinary cell renderer, the `value` prop is also available and it corresponds to the state of the row: `true` when expanded and `false` when collapsed.

Reviewer 2: ```suggestion
> **Note**: As any ordinary cell renderer, the `value` prop is also available and it corresponds to the state of the row: `true` when expanded and `false` when collapsed.
```
?

### Disable detail panel content scroll

> By default, the detail panel has as width the sum of the widths of all columns.

Reviewer 2: ```suggestion
By default, the detail panel has a width that is the sum of the widths of all columns.
```

> This means that, when a horizontal scrollbar is present, scrolling it will also scroll the panel content.

Reviewer 2: ```suggestion
This means that when a horizontal scrollbar is present, scrolling it will also scroll the panel content.
```

To avoid this behavior, set the size of the detail panel to the outer size of the grid.
Use `apiRef.current.getRootDimensions()` to get the latest dimension values.
Finally, to prevent the panel from scrolling, set `position: sticky` and `left: 0`.

> The following demo demonstrates how this can be achieved.

Reviewer 2: ```suggestion
The following demo shows how this can be achieved.
```

> Notice that the toggle column is pinned to make sure that it will always be visible when scrolled.

Reviewer 2: ```suggestion
Notice that the toggle column is pinned to make sure that it will always be visible when the grid is scrolled.
```

  getDetailPanelHeight={({ row }) => 100} // Optional, default is 500px.
/>
```

To expand a row click on the "i" icon, or press <kbd><kbd class="key">Ctrl</kbd>+<kbd class="key">Enter</kbd></kbd> inside one of the cells of the row.
Returning `null` or `undefined` as the value of `getDetailPanelContent` will prevent the respective row from being expanded.

{{"demo": "pages/components/data-grid/group-pivot/BasicDetailPanels.js", "bg": "inline", "defaultCodeOpen": false}}

> ⚠ Always memoize the function provided to `getDetailPanelContent` and `getDetailPanelHeight`.
> The grid depends on the referential value of these props to cache their values and optimize the rendering.
>
> ```tsx
> const getDetailPanelContent = React.useCallback(() => { ... }, []);
>
> <DataGridPro getDetailPanelContent={getDetailPanelContent} />
> ```
>
> ⚠ Depending on the height of the detail panel, you may see a blank space when scrolling.
> This is caused by the grid using a lazy approach to update the rendered rows.
> Set `rowThreshold` to 0 to force new rows to be rendered more often to fill the blank space.
> Note that this may reduce the performance.
>
> ```tsx
> <DataGridPro rowThreshold={0} />
> ```

> ### Controlling the detail panels

Author: I went with "Controlling expanded detail panels".

To control which rows are expanded, pass a list of row IDs to the `detailPanelExpandedRowIds` prop.
Passing a callback to the `onDetailPanelExpandedRowIds` prop can be used to detect when a row was expanded or collapsed.

> On the other hand, if you only want to initialize the grid with some rows already expanded, use the `initialState` prop as follow:

Reviewer 2: ```suggestion
On the other hand, if you only want to initialize the grid with some rows already expanded, use the `initialState` prop as follows:
```

> On the other hand, if you only want to initialize the grid with some rows already expanded, use the `initialState` prop as follow:

Author: I mentioned `initialState` only because we support it but I don't see a use case for it.
